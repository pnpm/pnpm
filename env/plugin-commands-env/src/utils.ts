import { promises as fs } from 'fs'
import path from 'path'

export const CURRENT_NODE_DIRNAME = 'nodejs_current'

export async function getNodeExecPathAndTargetDir (pnpmHomeDir: string): Promise<{ nodePath: string, nodeLink?: string }> {
  const nodePath = getNodeExecPathInBinDir(pnpmHomeDir)
  const nodeCurrentDirLink = path.join(pnpmHomeDir, CURRENT_NODE_DIRNAME)
  let nodeCurrentDir: string | undefined
  try {
    nodeCurrentDir = await fs.readlink(nodeCurrentDirLink)
  } catch (err) {
    nodeCurrentDir = undefined
  }
  return { nodePath, nodeLink: nodeCurrentDir ? getNodeExecPathInNodeDir(nodeCurrentDir) : undefined }
}

export function getNodeExecPathInBinDir (pnpmHomeDir: string): string {
  return path.resolve(pnpmHomeDir, process.platform === 'win32' ? 'node.exe' : 'node')
}

export function getNodeExecPathInNodeDir (nodeDir: string): string {
  return path.join(nodeDir, process.platform === 'win32' ? 'node.exe' : 'bin/node')
}
