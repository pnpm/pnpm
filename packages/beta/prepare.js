const fs = require('fs')
const path = require('path')

fs.writeFileSync(path.join(__dirname, 'pnpm'), '', 'utf8')
