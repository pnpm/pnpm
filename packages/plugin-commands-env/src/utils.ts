import { promises as fs } from 'fs'
import path from 'path'

export async function getNodeExecPathAndTargetDir (pnpmHomeDir: string) {
  const nodePath = getNodeExecPathInBinDir(pnpmHomeDir)
  let nodeLink: string | undefined
  try {
    nodeLink = await fs.readlink(nodePath)
  } catch (err) {
    nodeLink = undefined
  }
  return { nodePath, nodeLink }
}

export function getNodeExecPathInBinDir (pnpmHomeDir: string) {
  return path.resolve(pnpmHomeDir, process.platform === 'win32' ? 'node.exe' : 'node')
}

export function getNodeExecPathInNodeDir (nodeDir: string) {
  return path.join(nodeDir, process.platform === 'win32' ? 'node.exe' : 'bin/node')
}
