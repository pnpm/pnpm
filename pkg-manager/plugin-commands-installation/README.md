# @pnpm/plugin-commands-installation

> Commands for installation

[![npm version](https://img.shields.io/npm/v/@pnpm/plugin-commands-installation.svg)](https://www.npmjs.com/package/@pnpm/plugin-commands-installation)

## Installation

```sh
pnpm add @pnpm/plugin-commands-installation
```

## License

MIT
