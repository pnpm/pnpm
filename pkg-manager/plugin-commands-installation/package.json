{
  "name": "@pnpm/plugin-commands-installation",
  "version": "11.5.2",
  "description": "Commands for installation",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "!*.map"
  ],
  "engines": {
    "node": ">=14.6"
  },
  "scripts": {
    "start": "tsc --watch",
    "lint": "eslint \"src/**/*.ts\" \"test/**/*.ts\"",
    "registry-mock": "registry-mock",
    "test:jest": "jest",
    "test:e2e": "registry-mock prepare && run-p -r registry-mock test:jest",
    "_test": "cross-env PNPM_REGISTRY_MOCK_PORT=7775 pnpm run test:e2e",
    "test": "pnpm run compile && pnpm run _test",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/main/pkg-manager/plugin-commands-installation",
  "keywords": [
    "pnpm7",
    "pnpm"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "homepage": "https://github.com/pnpm/pnpm/blob/main/pkg-manager/plugin-commands-installation#readme",
  "devDependencies": {
    "@pnpm/assert-project": "workspace:*",
    "@pnpm/lockfile-types": "workspace:*",
    "@pnpm/modules-yaml": "workspace:*",
    "@pnpm/plugin-commands-installation": "workspace:*",
    "@pnpm/prepare": "workspace:*",
    "@pnpm/registry-mock": "3.5.0",
    "@pnpm/test-fixtures": "workspace:*",
    "@types/proxyquire": "^1.3.28",
    "@types/ramda": "0.28.20",
    "@types/sinon": "^10.0.13",
    "@types/yarnpkg__lockfile": "^1.1.5",
    "@types/zkochan__table": "npm:@types/table@6.0.0",
    "delay": "^5.0.0",
    "jest-diff": "^29.4.3",
    "path-name": "^1.0.0",
    "proxyquire": "^2.1.3",
    "read-yaml-file": "^2.1.0",
    "sinon": "^15.0.1",
    "symlink-dir": "^5.1.1",
    "tempy": "^1.0.1",
    "write-json-file": "^4.3.0",
    "write-pkg": "4.0.0",
    "write-yaml-file": "^4.2.0"
  },
  "dependencies": {
    "@pnpm/cli-utils": "workspace:*",
    "@pnpm/colorize-semver-diff": "^1.0.1",
    "@pnpm/command": "workspace:*",
    "@pnpm/common-cli-options-help": "workspace:*",
    "@pnpm/config": "workspace:*",
    "@pnpm/constants": "workspace:*",
    "@pnpm/core": "workspace:*",
    "@pnpm/error": "workspace:*",
    "@pnpm/filter-workspace-packages": "workspace:*",
    "@pnpm/find-workspace-dir": "workspace:*",
    "@pnpm/find-workspace-packages": "workspace:*",
    "@pnpm/graceful-fs": "workspace:*",
    "@pnpm/manifest-utils": "workspace:*",
    "@pnpm/matcher": "workspace:*",
    "@pnpm/outdated": "workspace:*",
    "@pnpm/package-store": "workspace:*",
    "@pnpm/parse-wanted-dependency": "workspace:*",
    "@pnpm/plugin-commands-rebuild": "workspace:*",
    "@pnpm/pnpmfile": "workspace:*",
    "@pnpm/read-project-manifest": "workspace:*",
    "@pnpm/resolver-base": "workspace:*",
    "@pnpm/semver-diff": "^1.1.0",
    "@pnpm/sort-packages": "workspace:*",
    "@pnpm/store-connection-manager": "workspace:*",
    "@pnpm/types": "workspace:*",
    "@pnpm/workspace.pkgs-graph": "workspace:*",
    "@yarnpkg/core": "4.0.0-rc.27",
    "@yarnpkg/lockfile": "^1.1.0",
    "@yarnpkg/parsers": "3.0.0-rc.27",
    "@zkochan/rimraf": "^2.1.2",
    "@zkochan/table": "^1.0.0",
    "@zkochan/which": "^2.0.3",
    "chalk": "^4.1.2",
    "ci-info": "^3.8.0",
    "enquirer": "^2.3.6",
    "is-subdir": "^1.2.0",
    "load-json-file": "^6.2.0",
    "mem": "^8.1.1",
    "p-filter": "^2.1.0",
    "p-limit": "^3.1.0",
    "path-absolute": "^1.0.1",
    "path-exists": "^4.0.0",
    "ramda": "npm:@pnpm/ramda@0.28.1",
    "render-help": "^1.0.3",
    "version-selector-type": "^3.0.0"
  },
  "peerDependencies": {
    "@pnpm/logger": "^5.0.0"
  },
  "funding": "https://opencollective.com/pnpm",
  "exports": {
    ".": "./lib/index.js"
  }
}
