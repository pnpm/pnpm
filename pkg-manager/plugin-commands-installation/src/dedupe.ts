import { docsUrl } from '@pnpm/cli-utils'
import { OPTIONS, UNIVERSAL_OPTIONS } from '@pnpm/common-cli-options-help'
import { dedupeDiffCheck } from '@pnpm/dedupe.check'
import renderHelp from 'render-help'
import { type InstallCommandOptions, rcOptionsTypes as installCommandRcOptionsTypes } from './install'
import { installDeps } from './installDeps'
import omit from 'ramda/src/omit'

// In general, the "pnpm dedupe" command should use .npmrc options that "pnpm install" would also accept.
export function rcOptionsTypes (): Record<string, unknown> {
  // Some options on pnpm install (like --frozen-lockfile) don't make sense on pnpm dedupe.
  return omit(['frozen-lockfile'], installCommandRcOptionsTypes())
}

export function cliOptionsTypes (): Record<string, unknown> {
  return {
    ...rcOptionsTypes(),
    check: Boolean,
  }
}

export const commandNames = ['dedupe']

export function help (): string {
  return renderHelp({
    description: 'Perform an install removing older dependencies in the lockfile if a newer version can be used.',
    descriptionLists: [
      {
        title: 'Options',
        list: [
          ...UNIVERSAL_OPTIONS,
          {
            description: 'Check if running dedupe would result in changes without installing packages or editing the lockfile. Exits with a non-zero status code if changes are possible.',
            name: '--check',
          },
          OPTIONS.ignoreScripts,
          OPTIONS.offline,
          OPTIONS.preferOffline,
          OPTIONS.storeDir,
          OPTIONS.virtualStoreDir,
          OPTIONS.globalDir,
        ],
      },
    ],
    url: docsUrl('dedupe'),
    usages: ['pnpm dedupe'],
  })
}

export interface DedupeCommandOptions extends InstallCommandOptions {
  readonly check?: boolean
}

export async function handler (opts: DedupeCommandOptions): Promise<void> {
  const include = {
    dependencies: opts.production !== false,
    devDependencies: opts.dev !== false,
    optionalDependencies: opts.optional !== false,
  }
  return installDeps({
    ...opts,
    dedupe: true,
    include,
    includeDirect: include,
    lockfileCheck: opts.check ? dedupeDiffCheck : undefined,
  }, [])
}
