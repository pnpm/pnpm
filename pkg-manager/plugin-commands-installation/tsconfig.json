{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts",
    "../../__typings__/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../../__utils__/assert-project"
    },
    {
      "path": "../../__utils__/prepare"
    },
    {
      "path": "../../__utils__/test-fixtures"
    },
    {
      "path": "../../cli/cli-utils"
    },
    {
      "path": "../../cli/command"
    },
    {
      "path": "../../cli/common-cli-options-help"
    },
    {
      "path": "../../config/config"
    },
    {
      "path": "../../config/matcher"
    },
    {
      "path": "../../dedupe/check"
    },
    {
      "path": "../../exec/plugin-commands-rebuild"
    },
    {
      "path": "../../fs/graceful-fs"
    },
    {
      "path": "../../hooks/pnpmfile"
    },
    {
      "path": "../../lockfile/lockfile-types"
    },
    {
      "path": "../../packages/constants"
    },
    {
      "path": "../../packages/error"
    },
    {
      "path": "../../packages/parse-wanted-dependency"
    },
    {
      "path": "../../packages/types"
    },
    {
      "path": "../../pkg-manifest/manifest-utils"
    },
    {
      "path": "../../pkg-manifest/read-project-manifest"
    },
    {
      "path": "../../resolving/resolver-base"
    },
    {
      "path": "../../reviewing/outdated"
    },
    {
      "path": "../../store/package-store"
    },
    {
      "path": "../../store/store-connection-manager"
    },
    {
      "path": "../../workspace/filter-workspace-packages"
    },
    {
      "path": "../../workspace/find-workspace-dir"
    },
    {
      "path": "../../workspace/find-workspace-packages"
    },
    {
      "path": "../../workspace/pkgs-graph"
    },
    {
      "path": "../../workspace/sort-packages"
    },
    {
      "path": "../core"
    },
    {
      "path": "../modules-yaml"
    }
  ],
  "composite": true
}
