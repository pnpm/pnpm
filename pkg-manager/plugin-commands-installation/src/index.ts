import * as add from './add'
import * as ci from './ci'
import * as dedupe from './dedupe'
import * as install from './install'
import * as fetch from './fetch'
import * as link from './link'
import * as prune from './prune'
import * as remove from './remove'
import * as unlink from './unlink'
import * as update from './update'
import * as importCommand from './import'

export { add, ci, dedupe, fetch, install, link, prune, remove, unlink, update, importCommand }
