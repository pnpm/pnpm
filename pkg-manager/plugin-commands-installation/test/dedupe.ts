import fs from 'fs'
import path from 'path'
import { DedupeCheckIssuesError } from '@pnpm/dedupe.check'
import { readProjects } from '@pnpm/filter-workspace-packages'
import { type Lockfile } from '@pnpm/lockfile-types'
import { dedupe, install } from '@pnpm/plugin-commands-installation'
import { prepare } from '@pnpm/prepare'
import { fixtures } from '@pnpm/test-fixtures'
import { createTestIpcServer } from '@pnpm/test-ipc-server'
import { diff } from 'jest-diff'
import readYamlFile from 'read-yaml-file'
import { DEFAULT_OPTS } from './utils'

const f = fixtures(__dirname)

describe('pnpm dedupe', () => {
  test('updates old resolutions from importers block and removes old packages', async () => {
    const { originalLockfile, dedupedLockfile, dedupeCheckError } = await testFixture('workspace-with-lockfile-dupes')
    // Many old packages should be deleted as result of deduping. See snapshot file for details.
    expect(diff(originalLockfile, dedupedLockfile, diffOptsForLockfile)).toMatchSnapshot()
    expect(dedupeCheckError.dedupeCheckIssues).toEqual({
      importerIssuesByImporterId: {
        added: [],
        removed: [],
        updated: {
          'packages/bar': {
            ajv: {
              next: '6.12.6',
              prev: '6.10.2',
              type: 'updated',
            },
          },
        },
      },
      packageIssuesByDepPath: {
        added: [],
        removed: [
          '/ajv/6.10.2',
          '/fast-deep-equal/2.0.1',
          '/fast-json-stable-stringify/2.0.0',
          '/punycode/2.1.1',
          '/uri-js/4.2.2',
        ],
        updated: {},
      },
    })
  })

  test('updates old resolutions from package block', async () => {
    const { originalLockfile, dedupedLockfile, dedupeCheckError } = await testFixture('workspace-with-lockfile-subdep-dupes')
    // This is a smaller scale test that should just update uri-js@4.2.2 to
    // punycode@2.3.0 and remove punycode@2.1.1. See snapshot file for details.
    expect(diff(originalLockfile, dedupedLockfile, diffOptsForLockfile)).toMatchSnapshot()
    expect(dedupeCheckError.dedupeCheckIssues).toEqual({
      importerIssuesByImporterId: {
        added: [],
        removed: [],
        updated: {},
      },
      packageIssuesByDepPath: {
        added: [],
        removed: [
          '/punycode/2.1.1',
        ],
        updated: {
          '/uri-js/4.2.2': {
            punycode: {
              next: '2.3.0',
              prev: '2.1.1',
              type: 'updated',
            },
          },
        },
      },
    })
  })

  test('dedupe: ignores all the lifecycle scripts when --ignore-scripts is used', async () => {
    await using server = await createTestIpcServer()

    const project = prepare({
      name: 'test-dedupe-with-ignore-scripts',
      version: '0.0.0',

      scripts: {
        // eslint-disable:object-literal-sort-keys
        preinstall: server.sendLineScript('preinstall'),
        prepare: server.sendLineScript('prepare'),
        postinstall: server.sendLineScript('postinstall'),
        // eslint-enable:object-literal-sort-keys
      },
    })

    const opts = {
      ...DEFAULT_OPTS,
      recursive: true,
      dir: project.dir(),
      ignoreScripts: true,
      lockfileDir: project.dir(),
      workspaceDir: project.dir(),
    }

    await install.handler(opts)

    await dedupe.handler(opts)

    expect(fs.existsSync('package.json')).toBeTruthy()
    expect(server.getLines()).toStrictEqual([])
  })
})

const noColor = (str: string) => str
const diffOptsForLockfile = {
  // Avoid showing common lines to make the snapshot smaller and less noisy.
  // https://github.com/facebook/jest/tree/05deb8393c4ad71/packages/jest-diff#example-of-options-to-limit-common-lines
  contextLines: 3,
  expand: false,

  // Remove color from snapshots
  // https://github.com/facebook/jest/tree/05deb8393c4ad71/packages/jest-diff#example-of-options-for-no-colors
  aColor: noColor,
  bColor: noColor,
  changeColor: noColor,
  commonColor: noColor,
  patchColor: noColor,
}

async function testFixture (fixtureName: string) {
  const project = prepare(undefined)
  f.copy(fixtureName, project.dir())

  const { allProjects, selectedProjectsGraph } = await readProjects(project.dir(), [])

  const opts = {
    ...DEFAULT_OPTS,
    allProjects,
    selectedProjectsGraph,
    recursive: true,
    dir: project.dir(),
    lockfileDir: project.dir(),
    workspaceDir: project.dir(),
    resolutionMode: 'highest' as const, // TODO: this should work with the default resolution mode (TODOv8)
  }

  const readProjectLockfile = () => readYamlFile<Lockfile>(path.join(project.dir(), './pnpm-lock.yaml'))

  const originalLockfile = await readProjectLockfile()

  // Sanity check that this test is set up correctly by ensuring the lockfile is
  // unmodified after a regular install.
  await install.handler(opts)
  expect(await readProjectLockfile()).toEqual(originalLockfile)

  let dedupeCheckError: DedupeCheckIssuesError | undefined
  try {
    await dedupe.handler({ ...opts, check: true })
  } catch (err: unknown) {
    expect(err).toBeInstanceOf(DedupeCheckIssuesError)
    dedupeCheckError = err as DedupeCheckIssuesError
  } finally {
    // The dedupe check option should never change the lockfile.
    expect(await readProjectLockfile()).toEqual(originalLockfile)
  }

  if (dedupeCheckError == null) {
    throw new Error('Expected change report from pnpm dedupe --check')
  }

  // The lockfile fixture has several packages that could be removed after
  // re-resolving versions.
  await dedupe.handler(opts)

  const dedupedLockfile = await readProjectLockfile()

  // It should be possible to remove packages from the fixture lockfile.
  const originalLockfilePackageNames = Object.keys(originalLockfile.packages ?? {})
  const dedupedLockfilePackageNames = Object.keys(dedupedLockfile.packages ?? {})
  expect(dedupedLockfilePackageNames.length).toBeLessThan(originalLockfilePackageNames.length)

  // The "pnpm dedupe" command should only remove packages when the lockfile is
  // up to date. Ensure no new packages/dependencies were added.
  expect(originalLockfilePackageNames).toEqual(expect.arrayContaining(dedupedLockfilePackageNames))

  // Run pnpm install one last time to ensure the deduped lockfile is in a good
  // state. If so, the "pnpm install" command should pass successfully and not
  // make any further edits to the lockfile.
  await install.handler(opts)
  expect(await readProjectLockfile()).toEqual(dedupedLockfile)

  return { originalLockfile, dedupedLockfile, dedupeCheckError }
}
