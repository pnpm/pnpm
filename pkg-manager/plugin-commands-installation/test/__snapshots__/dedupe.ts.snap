// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pnpm dedupe updates old resolutions from importers block and removes old packages 1`] = `
"- Expected
+ Received

@@ -5,7 +5,7 @@
        "dependencies": Object {
          "ajv": Object {
            "specifier": "^6.10.2",
-           "version": "6.10.2",
+           "version": "6.12.6",
          },
        },
      },
@@ -20,18 +20,6 @@
    },
    "lockfileVersion": "6.0",
    "packages": Object {
-     "/ajv@6.10.2": Object {
-       "dependencies": Object {
-         "fast-deep-equal": "2.0.1",
-         "fast-json-stable-stringify": "2.0.0",
-         "json-schema-traverse": "0.4.1",
-         "uri-js": "4.2.2",
-       },
-       "dev": false,
-       "resolution": Object {
-         "integrity": "sha512-TXtUUEYHuaTEbLZWIKUr5pmBuhDLy+8KYtPYdcV8qC+pOZL+NKqYwvWSRrVXHn+ZmRRAu8vJTAznH7Oag6RVRw==",
-       },
-     },
      "/ajv@6.12.6": Object {
        "dependencies": Object {
          "fast-deep-equal": "3.1.3",
@@ -42,24 +30,12 @@
        "dev": false,
        "resolution": Object {
          "integrity": "sha512-j3fVLgvTo527anyYyJOGTYJbG+vnnQYvE0m5mmkc1TK+nxAppkCLMIL0aZ4dblVCNoGShhm+kzE4ZUykBoMg4g==",
-       },
-     },
-     "/fast-deep-equal@2.0.1": Object {
-       "dev": false,
-       "resolution": Object {
-         "integrity": "sha512-bCK/2Z4zLidyB4ReuIsvALH6w31YfAQDmXMqMx6FyfHqvBxtjC0eRumeSu4Bs3XtXwpyIywtSTrVT99BxY1f9w==",
        },
      },
      "/fast-deep-equal@3.1.3": Object {
        "dev": false,
        "resolution": Object {
          "integrity": "sha512-f3qQ9oQy9j2AhBe/H9VC91wLmKBCCU/gDOnKNAYG5hswO7BLKj09Hc5HYNz9cGI++xlpDCIgDaitVs03ATR84Q==",
-       },
-     },
-     "/fast-json-stable-stringify@2.0.0": Object {
-       "dev": false,
-       "resolution": Object {
-         "integrity": "sha1-1RQsDK7msRifh9OnYREGT4bIu/I=",
        },
      },
      "/fast-json-stable-stringify@2.1.0": Object {
@@ -73,16 +49,7 @@
        "resolution": Object {
          "integrity": "sha512-xbbCH5dCYU5T8LcEhhuh7HJ88HXuW3qsI3Y0zOZFKfZEHcpWiHU/Jxzk629Brsab/mMiHQti9wMP+845RPe3Vg==",
        },
-     },
-     "/punycode@2.1.1": Object {
-       "dev": false,
-       "engines": Object {
-         "node": ">=6",
      },
-       "resolution": Object {
-         "integrity": "sha512-XRsRjdf+j5ml+y/6GKHPZbrF/8p2Yga0JPtdqTIY2Xe5ohJPD9saDJJLPvp9+NSBprVvevdXZybnj2cv8OEd0A==",
-       },
-     },
      "/punycode@2.3.0": Object {
        "dev": false,
        "engines": Object {
@@ -90,15 +57,6 @@
        },
        "resolution": Object {
          "integrity": "sha512-rRV+zQD8tVFys26lAGR9WUuS4iUAngJScM+ZRSKtvl5tKeZ2t5bvdNFdNHBW9FWR4guGHlgmsZ1G7BSm2wTbuA==",
-       },
-     },
-     "/uri-js@4.2.2": Object {
-       "dependencies": Object {
-         "punycode": "2.1.1",
-       },
-       "dev": false,
-       "resolution": Object {
-         "integrity": "sha512-KY9Frmirql91X2Qgjry0Wd4Y+YTdrdZheS8TFwvkbLWf/G5KNJDCh6pKL5OZctEW4+0Baa5idK2ZQuELRwPznQ==",
        },
      },
      "/uri-js@4.4.1": Object {"
`;

exports[`pnpm dedupe updates old resolutions from package block 1`] = `
"- Expected
+ Received

@@ -20,15 +20,6 @@
    },
    "lockfileVersion": "6.0",
    "packages": Object {
-     "/punycode@2.1.1": Object {
-       "dev": false,
-       "engines": Object {
-         "node": ">=6",
-       },
-       "resolution": Object {
-         "integrity": "sha512-XRsRjdf+j5ml+y/6GKHPZbrF/8p2Yga0JPtdqTIY2Xe5ohJPD9saDJJLPvp9+NSBprVvevdXZybnj2cv8OEd0A==",
-       },
-     },
      "/punycode@2.3.0": Object {
        "dev": false,
        "engines": Object {
@@ -40,7 +31,7 @@
      },
      "/uri-js@4.2.2": Object {
        "dependencies": Object {
-         "punycode": "2.1.1",
+         "punycode": "2.3.0",
        },
        "dev": false,
        "resolution": Object {"
`;
